feat(core): `BundlingDockerImage` now supports `run()` and `cp()` utilities

misterjoshua opened the pull request:
This PR introduces `run()` and `cp()` utility methods to `BundlingDockerImage`. After building their own Docker images, users can more easily run the image or copy files out of the image to create their own assets without using the bundling mechanism.

Possibly closes #9329

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: ee60c5ae051cec94a348f9648989c7f01317d27a
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=eb3ae789-edcc-41ca-b266-698bffe3afe5%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 5954cbd1dd4a5b3f0ddcb70122da9566ef71b185
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=cea08224-0f9f-4718-8b3a-f12e4e1b4d87%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: ffc60588ad618efde1157dbf3ce3a188476e9f53
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=ef3ddec3-b879-4716-b80f-a4c67c3b1cd0%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: df47ba8b31d4dda69b102d996168287139771867
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=c760fdbf-e743-4f59-a5e5-08016ef28296%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 81e08d243256bca5275b5d3bf04ba1071d53b8ae
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=4bb2b0e8-90a5-4319-be58-62a43d94e368%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 09764d3e14d5e8d346736b69276abe48cfbe6ca4
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=fc0cda7a-3a49-4efd-98ed-27990583bc94%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: dc75417bb3b425e0d4d7a20edeb0246387f6bc7d
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=f0e178c2-8504-4492-8039-b577d8a01024%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

jogold commented on packages/@aws-cdk/aws-lambda/lib/code.ts:
```diff
@@ -51,6 +51,24 @@ export abstract class Code {
   }
 
+  /**
+   * Provides a code asset by building a Dockerfile and running it. The docker
+   * image should copy its code to /asset-output when it is run.
```
This is really interesting. I'm wondering if there's something to do in `@aws-cdk/core` here instead.

If the asset is built by building a Dockerfile then why should we mount `/asset-input` or run the container? Volume mounts are heavy in terms of performance (see also issues with macOS).

If the asset is created at a well known location then we can do something like this after building the Dockerfile:

```bash
id=$(docker create image-name-from-build) # does not start the container
docker cp $id:/asset-output cdk.out/xxx # copy asset to staging dir
docker rm -v $id # remove container
```

misterjoshua replied:
That's a very good point! If we were to add a new member to `BundlingOptions`, like `imageStaticAssetPath` (or something like that), we could handle the docker create, cp, and rm from the `bundle` function in `AssetStaging`.

misterjoshua replied:
> That's a very good point! If we were to add a new member to `BundlingOptions`, like `imageStaticAssetPath` (or something like that), we could handle the docker create, cp, and rm from the `bundle` function in `AssetStaging`.
> 
> A reason for `imageStaticAssetPath` over any specific path would be the case where a user wants one docker build that emits several different assets - each bundler pass may use the same image, but would access different directories in the image. But, it makes a lot of sense to set a default path like `/asset-output`.

A new commit implementing this idea is now up.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 9638e64f85fd2292d106e853792970d0ac4fa4d7
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=b0898fe4-25bb-44f6-9960-a25055228932%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: f4a0b3eb625a918b2860216ee7f94440212e6b03
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=cfcca4ab-dfa3-4a85-aea3-19c8ff0047b8%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 2135c356cc8ed84f044012c7f941881b35f07cfb
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=ecca0585-2031-45ef-bfaf-b818e67d027c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

misterjoshua commented:
New changes to reflect the following:
* Changed the bundler option name to `copyFromImagePath`.  The former is a better option name than `imageStaticAssetPath` because it better conveys what the bundler will do.
* Removed `lambda.Code.fromDockerBuild` as `lambda.Code.fromAsset` supports the new bundler option out of the box.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 8148472a6858843155c5d31fd839cdaccc7446c3
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=3f778be8-547a-47a6-949d-737741388820%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 7b6318833230a61bd9e93509a9eedef708ae1a1d
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=aa8903d1-fe77-4e76-8daf-6280d5233c0b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: a7680ad4dfdecaed42a9ba18b7b9362b3ae40f17
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=87e86ff8-6115-4d40-97ab-161923a22648%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 64f9a97ad3ef2e2d824d332254349d93f6a5ffc4
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=86d7978d-8284-4a36-8315-3355d1ec39ea%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 61116f1adb5c96525f39ceea76b0b5faf0515784
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=41d00f0d-0a07-40cf-a28f-7837b0edb8a5%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: c0a7228b1b681dfcebd0c30e4517234c30dd3434
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=ecfbdbc4-1e30-48b7-967f-03ff645450e1%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 85971fcd8117e388b52c464b41db475616d7d2e7
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=b77b2596-dded-428b-8295-a6431dea51ee%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

nija-at commented on packages/@aws-cdk/aws-lambda/test/integ.from-docker-build.complex.ts:
```diff
@@ -0,0 +1,59 @@
+    });
+
+    const layer = new lambda.LayerVersion(this, 'Layer', {
+      code: lambda.Code.fromAsset(assetPath, {
```
Thanks @misterjoshua. I've assigned this to @eladb who is familiar with this part of the code base and will help you take this forward.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 77c5e80cca788adaacf44952acad7e886b796d3b
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=cf3cd0f8-7003-4012-b90a-73eb1ce9cc54%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

eladb commented:
@jogold could spare some cycles to work with @misterjoshua on this?

@misterjoshua something with the mental model is not intuitive to me at first glance. I am not sure I fully understand the use case and then the terminology of `copyFromImagePath` doesn't intuitively represent neither intent (the user's final goal) nor mechanism (what's mechanically happening).

misterjoshua commented:
FYI, I plan to fix the conflict and resume work on this PR as soon as I can, but I'm swamped this week.

eladb commented:
@jogold can you take a look?

jogold commented on packages/@aws-cdk/aws-s3-assets/README.md:
```diff
@@ -117,6 +119,32 @@ new assets.Asset(this, 'BundledAsset', {
 
+### Copying Directly From a Bundling Image
+
+You may set `copyFromImagePath` to `true` to instruct the bundler to copy an
+asset out of your bundling docker image. To use this feature, you should create
+your own bundling docker image from a `Dockerfile`:
```
When would you absolutely recommend using this? I can see the benefit but I find it hard to determine and clearly explain which use case should use `docker run` and which should use `docker build + cp`?

For `aws-lambda-nodejs` do we agree that it's not a good option because it would mean copying the whole `node_modules` folder in the image?

misterjoshua replied:
I would recommend using `docker build + cp` when the emitted asset consists of many small files. This is because copying thousands of files to the output volume in Docker for Windows is still slow with WSL 2. I'm not sure what is the case in macOS, but I suspect it's a similar situation.

My use case is a construct for a static site generator for my team. For one of my sites, the bundling/build results in nearly 40k files that I need to put on S3 behind CloudFront. Copying them from the bundler to the volume can take a minute longer than copying from the container with the docker command.

There's nothing stopping me from orchestrating the build outside of the bundler and creating the asset directly from those files. But, it would be more convenient if I could use the bundler for all the pre-existing docker-handling code.

Agreed that `aws-lambda-nodejs` may not benefit directly from this approach. Parcel is emitting a single file, right?

jogold commented on packages/@aws-cdk/aws-s3-assets/lib/asset.ts:
```diff
@@ -198,7 +198,17 @@ export class Asset extends cdk.Construct implements cdk.IAsset {
+ * from the bundler image, so the packaging should be a ZIP_DIRECTORY.
+ */
+function determinePackaging(assetPath: string, copyFromImagePath?: boolean): cdk.FileAssetPackaging {
+  if (copyFromImagePath) {
+    return cdk.FileAssetPackaging.ZIP_DIRECTORY;
+  }
```
This is a "trick" that shows that `path` has no meaning anymore here and to avoid checking its existence we need to return early.

misterjoshua commented on packages/@aws-cdk/aws-s3-assets/README.md:
````diff
@@ -117,6 +119,32 @@ new assets.Asset(this, 'BundledAsset', {
+
+```ts
+new assets.Asset(this, 'BundledAsset', {
+  path: '/asset-output', // path within the image to the asset
````
On reflection, I agree with you that using `path` like this doesn't fit well with the current Asset paradigm. You may be right that this is another type of asset if [my use case (below)](https://github.com/aws/aws-cdk/pull/9728#discussion_r484564510) is relevant to the CDK project.

jogold replied:
@eladb what do you thinK?

eladb commented:
What's the status here?

misterjoshua commented:
> What's the status here?

@eladb I believe @jogold requested [some feedback on whether we should consider implementing a different asset type](https://github.com/aws/aws-cdk/pull/9728#pullrequestreview-483573596).

eladb commented:
> > What's the status here?
> 
> @eladb I believe @jogold requested [some feedback](https://github.com/aws/aws-cdk/pull/9728#pullrequestreview-483573596) on whether we should consider implementing a different asset type. Overall, I'm not sure whether supporting my use case is out of scope for CDK bundling or the project as a whole, considering that I can handle this in a stage that happens before CDK runs.

You can also just write some code in your CDK app that prepares the asset directory (in whatever way) and then defines an asset that refers to this directory (no need to use bundling in this case).

What's the benefit of using the asset bundling mechanism? Is it just to run `docker`? Can we extract some of that logic into e.g. utility functions maybe?

misterjoshua commented:
> > > What's the status here?
> > 
> > 
> > @eladb I believe @jogold requested [some feedback](https://github.com/aws/aws-cdk/pull/9728#pullrequestreview-483573596) on whether we should consider implementing a different asset type. Overall, I'm not sure whether supporting my use case is out of scope for CDK bundling or the project as a whole, considering that I can handle this in a stage that happens before CDK runs.
> 
> You can also just write some code in your CDK app that prepares the asset directory (in whatever way) and then defines an asset that refers to this directory (no need to use bundling in this case).
> 
> What's the benefit of using the asset bundling mechanism? Is it just to run `docker`? Can we extract some of that logic into e.g. utility functions maybe?

@eladb Yes, I'd say the benefit lies in the ease of controlling docker to prepare such a directory. `BundlingDockerImage` is particularly convenient for building the image. And yes, I suspect that we can extract the docker logic into some utility functions to make it easier.

Does an API like this make sense?

```ts
// Build a Docker image
const image = BundlingDockerImage.fromAsset(contextPath, {
  // Added in this PR, lets you override the Dockerfile path - very useful for me
  file: dockerfilePath,
});

// My use case - copy files out of the image from `imagePath` to `localPath` on this computer via `docker cp`
image.copyOut(imagePath, localPath);
// Or alternatively, run the docker image with options.
image.run(dockerRunOptions);
```

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 8485802941a9135c418255a83e99e71587a38bea
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=0eb5cc16-1095-4ba1-9368-1c3293c4a194%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

eladb commented on packages/@aws-cdk/core/lib/bundling.ts:
```diff
@@ -110,6 +110,7 @@ export class BundlingDockerImage {
 
     const dockerArgs: string[] = [
       'build', '-q',
+      ...(options.file ? ['-f', options.file] : []),
```
I vaguely recall this is already supported. Did you rebase?

misterjoshua replied:
@eladb I wiped out my previous branch and started fresh from the master branch today - the old commits were not practical to revert. I used the force to push it over the branch in this PR. I didn't see pre-existing support for this [in master](https://github.com/aws/aws-cdk/blob/451200e3dba6b8d80bc8396dc0b7d4d84f29cb56/packages/%40aws-cdk/core/lib/bundling.ts#L111-L115). However, I've got practically the same code in PR #9582, which I believe you've previously reviewed.

jogold replied:
It's currently supported in `cdk-assets`.

misterjoshua replied:
The API of the `Docker` class in `cdk-assets` looks pretty handy to use. But, both `core` and `cdk-assets` seem to have minimal dependencies. Any ideas what I should do here?

jogold replied:
Nothing at this stage, I was just mentioning from where @eladb might have thought that this was already supported.

Not for this PR but at some point it might be interesting to correctly define the "roles" of `core` vs `cdk-assets`. I think that this was already discussed in another issue about where zipping should occur.

eladb commented on packages/@aws-cdk/core/lib/bundling.ts:
```diff
@@ -291,3 +304,21 @@ function dockerExec(args: string[], options?: SpawnSyncOptions) {
   return proc;
 }
+
+export function dockerCopyFromImage(image: string, imagePath: string, outputPath: string) {
```
Why isn't this a method?

misterjoshua replied:
I've now refactored this code into `BundlingDockerImage.copyOut`.

eladb commented on packages/@aws-cdk/core/lib/bundling.ts:
```diff
@@ -110,6 +110,7 @@ export class BundlingDockerImage {
 
```
@jogold does it make sense to rename this to `DockerImage`? What's "bundling" about this?

jogold replied:
It's the docker image used when bundling an asset/accepted in the `bundling` options... but I agree that there's nothing specific to bundling in its implementation.

eladb replied:
Let's leave this for a future change

eladb reviewed (dismissed):
I like this direction. See some comments

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: e002ef7d3f239963d6361dcf91c4240d49d5b3ec
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=b837c0f2-af47-4cdf-9c9a-8a6617751b0c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

misterjoshua commented:
@eladb I've pushed up changes from your review. It's time to sleep, so I'll check in later. :)

eladb commented on packages/@aws-cdk/core/lib/bundling.ts:
```diff
@@ -175,6 +174,27 @@ export class BundlingDockerImage {
+  /**
+   * Copies a file or directory out of the Docker image to the local filesystem
+   */
+  public copyOut(imagePath: string, outputPath: string) {
```
Should we rename to `cp` so it's aligned with the `docker` API?

misterjoshua replied:
Ok, I've pushed up this change.

eladb reviewed (commented):
Minor

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: f18d976b5fe686167bbca1dbebd18454071e3d89
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=92845a91-8405-4f2f-bda6-82c4b57a04b4%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: fa539cb9e661a877b2abe46246e700586aa4c930
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d6fcc4e0-8308-43da-94bd-a2e9a0078d38%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: c03391a4d8acb1c6fcc517438be2a7ae1a792a77
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d20b8305-885f-46ff-842d-b1fe577b79a6%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: f0224c2152aeae2b5dd5007b294e0e0c6dc781e1
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=5dbf6cf2-0b9f-4ac0-a2c2-29db3c1b1a3f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

misterjoshua commented:
@eladb I've made your `cp()` change and updated from the master branch. Is there anything else you'd like to see added?

misterjoshua commented:
I see some merge conflicts have arisen. I'll see what's required and push up a commit.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: f50ce4780ef03bd9f460ad9f323d3fac69e89ae6
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=ec4a8dde-6674-4305-9e6b-afc43eba2a3f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

misterjoshua commented:
@eladb I resolved a conflict due to the new `nodeunitShim`, so I think this may need another review. Though the fix was basically a parenthesis.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: c35b0f7dd4730b0c12a9079458aa6beca2c8523a
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=b24c384d-383d-4ac5-9297-50207e6a06eb%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
